### PR TITLE
orca: switching to stubserver in tests instead of testservice implementation

### PR DESCRIPTION
**Pull request by janardhanvissa**
Partially Addresses: https://github.com/grpc/grpc-go/issues/7291

RELEASE NOTES: None

### CODE REVIEW

File: orca/service_test.go [hunk 1]
```diff
@@ -112,11 +110,11 @@ func (s) TestE2E_CustomBackendMetrics_OutOfBand(t *testing.T) {
...
 	defer cc.Close()
 
-	// Spawn a goroutine which sends 20 unary RPCs to the test server. This
+	// Spawn a goroutine which sends 20 unary RPCs to the stub server. This
 	// will trigger the injection of custom backend metrics from the
-	// testServiceImpl.
+	// stubServer.
 	const numRequests = 20
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (purnesh42H) [inline orca/service_test.go:117]: Why do we need to change timeout here?

File: orca/service_test.go [hunk 2]
```diff
@@ -138,16 +136,17 @@ func (s) TestE2E_CustomBackendMetrics_OutOfBand(t *testing.T) {
 		t.Fatalf("Failed to create a stream for out-of-band metrics")
 	}
 
+	// Wait for the goroutine to finish before processing metrics.
+	if err := <-errCh; err != nil {
```
Comment (purnesh42H) [inline orca/service_test.go:140]: why is this out of the select-context block?

File: orca/service_test.go (hunk 1 above)
Comment (janardhanvissa) [reply orca/service_test.go]: The default is 5 secs, getting some context deadline issues. So, updated to 10 secs.

File: orca/service_test.go [hunk 3]
```diff
@@ -131,6 +130,8 @@ func (s) TestE2E_CustomBackendMetrics_OutOfBand(t *testing.T) {
 		errCh <- nil
 	}()
 
+	wg.Wait()
```
Comment (purnesh42H) [inline orca/service_test.go:133]: We don't need this waitGroup, there is already a for loop which Wait for the server to push metrics which indicate the completion of all the unary RPCs made from the above goroutine.

File: orca/service_test.go [hunk 4]
```diff
@@ -151,11 +147,15 @@ func (s) TestE2E_CustomBackendMetrics_OutOfBand(t *testing.T) {
 		default:
 		}
 
+		mu.Lock()
```
Comment (purnesh42H) [inline orca/service_test.go:150]: why did you add this? Please provide inline comments explaining such changes. The previous wantProto should work as is I think. If not, then please mention what is happening

File: orca/service_test.go [hunk 5]
```diff
@@ -151,11 +147,15 @@ func (s) TestE2E_CustomBackendMetrics_OutOfBand(t *testing.T) {
...
+		expectedUtilization := float64(requests) * 0.01
```
Comment (purnesh42H) [inline orca/service_test.go:151]: why did you change this value to float64(requests) * 0.01? it was `numRequests * 0.01`. The numRequests is a constant so that's we should have in want. requests are being calculated within stub server call and may not be synchronized. You shouldn't need mutex to assign a constant value

**Review by purnesh42H — APPROVED** [review]
lgtm

**Comment by purnesh42H** [comment]
@easwars for second review

File: orca/service_test.go [hunk 6]
```diff
@@ -93,14 +63,42 @@ func (s) TestE2E_CustomBackendMetrics_OutOfBand(t *testing.T) {
...
 		t.Fatalf("orca.EnableOutOfBandMetricsReportingForTesting() failed: %v", err)
 	}
 
-	// Register the test service implementation on the same grpc server, and start serving.
-	testgrpc.RegisterTestServiceServer(s, &testServiceImpl{smr: smr})
+	// Assign the gRPC server to the stub server and start serving.
+	stub.S = s
+	stubserver.StartTestService(t, stub)
 	go s.Serve(lis)
 	defer s.Stop()
```
Comment (arvindbr8) [inline orca/service_test.go:103]: Similar to my comment from the other PR. We can just pass in sopts to `StartTestService`

File: orca/service_test.go [hunk 7]
```diff
@@ -93,14 +63,42 @@ func (s) TestE2E_CustomBackendMetrics_OutOfBand(t *testing.T) {
...
 	if err := orca.Register(s, opts); err != nil {
 		t.Fatalf("orca.EnableOutOfBandMetricsReportingForTesting() failed: %v", err)
 	}
 
-	// Register the test service implementation on the same grpc server, and start serving.
-	testgrpc.RegisterTestServiceServer(s, &testServiceImpl{smr: smr})
+	// Assign the gRPC server to the stub server and start serving.
+	stub.S = s
+	stubserver.StartTestService(t, stub)
 	go s.Serve(lis)
```
Comment (arvindbr8) [inline orca/service_test.go:102]: StartTestService already does s.Serve. What is the expected behavior here

**Review by arvindbr8 — CHANGES_REQUESTED** [review]

**Comment by purnesh42H** [comment]
@arvindbr8 for re-review

File: orca/service_test.go [hunk 8]
```diff
@@ -93,16 +63,42 @@ func (s) TestE2E_CustomBackendMetrics_OutOfBand(t *testing.T) {
 	opts := orca.ServiceOptions{MinReportingInterval: shortReportingInterval, ServerMetricsProvider: smr}
 	internal.AllowAnyMinReportingInterval.(func(*orca.ServiceOptions))(&opts)
 
+	var requests int64
+	var mu sync.Mutex
```
Comment (arvindbr8) [inline orca/service_test.go:67]: can we consider using `atomics` here?

File: orca/service_test.go [hunk 9]
```diff
@@ -93,16 +63,42 @@ func (s) TestE2E_CustomBackendMetrics_OutOfBand(t *testing.T) {
...
+
+	const numRequests = 20
```
Comment (arvindbr8) [inline orca/service_test.go:69]: Let's move this back to where the usage is.

File: orca/service_test.go [hunk 10]
```diff
@@ -93,16 +63,42 @@ func (s) TestE2E_CustomBackendMetrics_OutOfBand(t *testing.T) {
...
+			smr.DeleteNamedUtilization(requestsMetricKey)
+			smr.SetCPUUtilization(0)
+			smr.SetMemoryUtilization(0)
+			smr.DeleteApplicationUtilization()
+			return &testpb.Empty{}, nil
+		},
+	}
+
+	// Assign the gRPC server to the stub server and start serving.
+	stub.Listener = lis
```
Comment (arvindbr8) [inline orca/service_test.go:94]: Let's specify this where `stub` is initialized

File: orca/service_test.go [hunk 11]
```diff
@@ -93,16 +63,42 @@ func (s) TestE2E_CustomBackendMetrics_OutOfBand(t *testing.T) {
...
+	var mu sync.Mutex
+
+	const numRequests = 20
+
+	stub := &stubserver.StubServer{
+		UnaryCallF: func(ctx context.Context, req *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
+			mu.Lock()
+			requests++
+
+			smr.SetNamedUtilization(requestsMetricKey, float64(requests)*0.01)
```
Comment (arvindbr8) [inline orca/service_test.go:76]: seems like we have changed how the lock is handled here, but this seems to be the right way of doing it. 

But this would go away if we use an atominc.Int64 here.

**Review by arvindbr8 — CHANGES_REQUESTED** [review]

File: orca/service_test.go [hunk 12]
```diff
@@ -93,16 +63,37 @@ func (s) TestE2E_CustomBackendMetrics_OutOfBand(t *testing.T) {
 	opts := orca.ServiceOptions{MinReportingInterval: shortReportingInterval, ServerMetricsProvider: smr}
 	internal.AllowAnyMinReportingInterval.(func(*orca.ServiceOptions))(&opts)
 
+	var requests atomic.Int64
+
+	stub := &stubserver.StubServer{
+		Listener: lis,
+		UnaryCallF: func(ctx context.Context, req *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
+
```
Comment (arvindbr8) [inline orca/service_test.go:71]:
```suggestion
```

**Review by arvindbr8 — APPROVED** [review]
LGTM
